android/src/main/java/io/ably/flutter/plugin/util/NotificationUtilities.java

```diff
@@ -0,0 +1,191 @@
+  // This method was adapted from [firebase-android-sdk's CommonNotificationBuilder.java](https://github.com/firebase/firebase-android-sdk/blob/master/firebase-messaging/src/main/java/com/google/firebase/messaging/CommonNotificationBuilder.java).
+  private static String getOrCreateChannel(Context context, String channelId) {
+    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
+      return null;
+    }
+    Bundle manifestMetadata = getManifestMetadata(context.getPackageManager(), context.getPackageName());
+    NotificationManager notificationManager = context.getSystemService(NotificationManager.class);
+
+    // If the channel already exists, just return the channelID that was passed in.
+    if (!TextUtils.isEmpty(channelId)) {
+      if (notificationManager.getNotificationChannel(channelId) != null) {
+        return channelId;
+      } else {
+        Log.w(
+            TAG,
+            "Notification Channel requested ("
+                + channelId
+                + ") has not been created by the app."
+                + " Manifest configuration, or default, value will be used.");
+      }
+    }
+
+    String manifestChannel = manifestMetadata.getString(METADATA_DEFAULT_CHANNEL_ID);
+    if (!TextUtils.isEmpty(manifestChannel)) {
+      if (notificationManager.getNotificationChannel(manifestChannel) != null) {
+        return manifestChannel;
+      } else {
+        Log.w(
+            TAG,
+            "Notification Channel set in AndroidManifest.xml has not been"
+                + " created by the app. Default value will be used.");
+      }
+    } else {
+      Log.w(
+          TAG,
+          "Missing Default Notification Channel metadata in AndroidManifest."
+              + " Default value will be used.");
+    }
+
+    // Create the default channel if it has not been created yet.
+    if (notificationManager.getNotificationChannel(FCM_FALLBACK_NOTIFICATION_CHANNEL) == null) {
+//       The Firebase SDK reads the "user visible name of the channel" from its string.xml using the code below.
+//          int channelLabelResourceId = context.getResources().getIdentifier(
+//                  FCM_FALLBACK_NOTIFICATION_CHANNEL_LABEL, "string", context.getPackageName());
+//      String channelName = context.getString(channelLabelResourceId);
+//      Instead, we just hardcode this as "Miscellaneous" below:
+
+      notificationManager.createNotificationChannel(
+          new NotificationChannel(
+              // channel id
+              FCM_FALLBACK_NOTIFICATION_CHANNEL,
+              // user visible name of the channel
+              "Miscellaneous",
+              // shows everywhere, makes noise, but does not visually intrude.
+              NotificationManager.IMPORTANCE_DEFAULT));
+    }
+
+    return FCM_FALLBACK_NOTIFICATION_CHANNEL;
+  }
+
+  private static Bundle getManifestMetadata(PackageManager packageManager, String packageName) {
```

Comment:
This one also should move to another class. Not directly related to notifications

Reply:
Most of this file contents is taken from Firebase, and I wanted to put in the same file. It's a private method.

Reply:
Repeating a comment from https://github.com/ably/ably-flutter/pull/166#discussion_r716573968:
> This file was deleted, because the entire functionality of this file IMHO should not be in the project. The functionality: Firebase Android SDK does not show notifications when the app is in the foreground, but on iOS it is customisable. I copied some code from Firebase and modified it to allow users to configure it.
>
> I have removed this now: c5070b3